### DISCUSSION TITLE
Dedicated Server?

### DISCUSSION THREAD

**Discussion by Baldie-dev** [NebulaModTeam/nebula#51] [Q&A] [answered]
Hi guys,

I hope this is not a stupid question. I saw this awesome mod and I would like to help and contribute. I saw that in the past there was **NebulaServer** project which ran in the console. Currently, the server is integrated into the game and it is implemented inside the Unity loop as MonoBehaviour.

My question is, are you testing it by running the server always from the game or do you have any workaround to run in dedicated? I have created my own NebulaServer project and I am trying to run it in the console. So I wonder if I am spending time to do something that was already solved :D thanks!

**Comment by hubastard** [comment, marked as answer]
We have moved from a dedicated server to a host / client architecture to avoid having to send the game state from the master client to the server and back to the remote clients.

There is a lot of data to synchronize and this was adding to much delay. With a host / client approach we have full game state access from the host and we can more easily send the state directly to remote clients and avoid additional round-trip time.

**Comment by Mushroom** [comment]
We moved away from the dedicated server model, as it would require the cloning of substantial parts of the game logic which is both error prone, and also a massive license violation.

It is much easier to just do a dedicated simulation on one host machine and send out authoritative updates to clients like we do now (or intend to at least). This is how many games do it, so it is not unusual (Stardew Valley, for example).

**Comment by Baldie-dev** [comment]
Thanks a lot for the answers, that make sense. I will have to check what are the limitations of the license. I had in my mind a binary that would load the game's DLL and basically run the already implemented game loop to do the simulation, it could also leverage the already implemented server. But maybe it is too much crazy :D

**Reply by Baldie-dev** [reply]
So it seems to be not possible in the end. If the Unity project is not compiled in the headless mode I would have to rewrite big portion of Unity Player :D Nah, too much work :D

**Reply by Baldie-dev** [reply]
Just found out that UnityPlayer offers the following argument: 
-batchmode | Run the game in “headless” mode. The game does not display anything or accept user input.
So maybe there would be some workaround using this.

**Reply by Baldie-dev** [reply]
seems to be working, the game is running in the background, and 0% utilization of GPU. Just the music and sounds are still playing. I will try to add support of reading additional parameters to immediately run the server from the game menu.

**Reply by Mushroom** [reply]
That would be useful, I have seen Stardew Valley mods that do something similar.

It would he interesting to see if things work properly due to the extensive use of the particles system and the GPU to offload stuff.
